### ISSUE TITLE
Support ??= as default value operator

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/language#3306]
Hello!

Currently when using default value, we do:

```dart
MyClass({this.number = 42});
```

This works by relying on whether the parameter was passed or not.
But a common alternative is to not rely on whether is passed or not, but whether it is null or not:

```dart
class MyClass {
  MyClass({int? number})
    : number ??= 42;

  final int number;
}
```

This can be quite tedious to use in part because it doesn't work with `this.value`/`super.value`, so we can't infer the type and have to type much more than needed.

Would it be possible to simplify this pattern with:

```dart
class MyClass {
  MyClass({this.number ??= 42});

  final int number;
}
```

Of course, the default value could be non-constant in this case. And the `number` parameter would be made nullable even if the property associated is non-nullable.

**Comment by jakemac53** [comment]
I generally like this idea a lot - it also gives us a general purpose mechanism for non-constant "default values". I think it is likely that we will want to talk about these as a different construct entirely from default values though, and instead talk about them as purely syntactic sugar.

It would need to be different for initializing formal parameters versus other parameters, which is a bit weird, but probably explainable (and most people won't actually care/notice).

Specifically, for an initializing formal parameter it would desugar to a normal parameter + `??=` initializer. The default type would be the nullable version of the field type:

```dart
class MyClass {
  final int number;
   
  // Original
  MyClass({this.number ??= 42});

  // Desugared
  MyClass({int? number}) : this.number = number ??= 42;
}
```

And then for regular parameters it would desugar to adding a `??=` assignment at the top of the body. I think we wouldn't want to implicitly change the parameter type, it should be explicitly declared as nullable but type promotion from the desugaring would make it non-nullable in the rest of the body:

```dart
class MyClass {
   
  // Original
  int myFunction({int? number ??= 42}) {
    return number; // allowed, it's promoted here!
  }

  // Desugared
  int myFunction({int? number}) {
    number ??= 42;
    return number; // allowed, it's promoted here!
  }
}
```

These assignments would be added in the order they appear in the function signature, and we could allow access to the other parameters within the assignments as well. These would also be type promoted where possible:

 ```dart
class MyClass {
  // Original
  int myFunction({
    int? a ??= b + 1, // Error, `b` is still nullable here
    int? b ??= 42,
    int? c ??= b + 1, // Fine, `b` is promoted here
  }) {
    ...
  }

  // Desugared
  int myFunction({int? number}) {
    number ??= 42;
    return number; // allowed, it's promoted here!
  }
}
```

This ability to get promotion within these expressions would not apply to initializing formal parameters though, since promotion would not take effect in the normal process of desugaring. That might be a bit hard to explain. 

But, in general this could be a frontend (and analyzer) only desugaring with ~0 work for the backends?

**Comment by lrhn** [comment]
I like that the ability to have non-constant default value expressions, which can have side effects, is limited to parameters that can't tell the difference between passing nothing and passing `null`.  That way we don't leave a way to do a "was this parameter passed or not" trapdoor. That also means that it still can't be used to implement `copyWith`.

Evaluation order is a tricky issue with non-constant expressions.

I'd say "as soon as possible" and "in source order", which would be *first thing in the body* for normal functions, as @jakemac53 says.
For constructors, it depends on the kind of constructor.
* Non-redirecting factory constructors: ... are basically just static functions.
* Redirecting factory constructors: Cannot have default values, and shouldn't have non-constant ones either.
* Redirecting generative constructors: Happens before the redirection argument expressions are evaluated.
* Non-redirecting generative constructors: Happens before the initializer list is evaluated (so references to an `int? this.x ??= 0` initializing formal will see the `0`, not the `null`. Which is final and has static type `int?` promoted to `int`, which is as close to `int` as you can get without being there).

I think that works. Most likely, we'll say that we do the evaluation *during* the binding-arguments-to-parameters phase, so the value actually bound to the *parameter* is the `argument-value ?? default-expression`. That also explains what happens with `final int? x ??= 0` when passed a `null` argument - the argument value is `null`, the value bound to the *`final`* parameter is `null ?? 0`, and that never changes. 
(And if the expression *is* constant, it can be used in a `const` constructor too.)

Another issue is how it interacts with variable type and finality.

**Comment by jakemac53** [comment]
> `final int? x ??= 0`

Ah yeah, if we care to support this (does _anybody_ ever actually use final parameters?), then it couldn't be just a simple desugaring as I proposed.

**Comment by abitofevrything** [comment]
Wouldn't it then be unnecessary for the parameter type to be nullable? E.g it'd be great if we could allow this:
```dart
void fn({int a ??= 5}) { ... }
```
Just as we currently allow
```dart
void fn({int a = 5}) { ... }
```

This would allow providing non constant default values to a parameter while still not allowing null to be passed as an argument. As it is currently (if it simply desugars to `a ??= 5`) then `a` should always be promoted to `int` anyway.

If we do implement it like this, it does however make code like this potentially confusing:
```dart
void fn({int? a ??= 5}) { ... }

void main() => fn(a: null);
```

In this case, since we passed `null`, the argument's default value wouldn't be used. This is however confusing with the fact that the `??=` operator currently checks for `null`, rather than a parameter that wasn't passed.

**Comment by abitofevrything** [comment]
An alternative would be to allow both `=` and `??=` on a parameter, like so:
```void fn({
  int? a = 5 ??= defaultValue
}) { ... }
```

The value of `a` is `5` if it was not passed, `defaultValue` if `null` was passed (or if no constant default value was provided), and the passed value otherwise. In either case, it's promoted to `int`.

**Comment by rrousselGit** [comment]
Indeed, it certainly makes function composition much simpler.
Default values relying on whether a parameter is passed or not are frustrating when making wrapper objects.

I've seen objects with a nullable parameter but a non-null default value. So when writing:

```dart
class MyWrapperWidget {
  MyWrapperWidget({this.parameter});
  final SomeType? parameter;

  build(context) {
    return WrappedWidget(parameter: parameter);
  }
}
```

The code _looks_ fine. But it behaves differently from the WrappedWidget, which uses a default value :shrug:

**Comment by jakemac53** [comment]
Yeah, a nullable parameter with a default value is almost certainly wrong. Do we have a lint for that?
